### DISCUSSION TITLE
Опыт использования

### DISCUSSION THREAD

**Discussion by Diman-kos** [djpont/php-tochka-bank-telegram-notifications#1] [General]
Добрый день!

Спасибо за Вашу работу!
Есть пара вопросов/замечаний )

1. После настройки конфига при первом запуске выводится ссылка "Авторизоваться". При нажатии ничего не происходило.
Проанализировав код добавил ?auth. Авторизация приложения в точке прошла успешно!

2. При каждом запуске index.php скрипт пересоздаёт хук заново:

Авторизация прошла успешно.
Проверяю вебхук.
Вебхук треюуется обновить.
Удаляю вебхук...
Создаю новый вебхук...
Новый вебхук создан!

Мне кажется так не должно быть. Или я не прав?

**Comment by djpont** [comment]
Здравствуйте!
Приятно видеть интерес к моему репозиторию.
1. Проверю, возможно действительно забыл ?auth.
2. Метод *checkWebHooks* проверяет наличие вебхука. По идее, при повторном
открытии странички вебхук не должен пересоздаваться. Но при повторной
авторизации будет повторно добавлять.

А вы проверили, уведомления о платежах приходят?
Сам класс *TochkaBank* рабочий, проверено на реальном проекте, уведомления
приходят.

С уважением, Вотяков Александр.


сб, 1 июл. 2023 г. в 15:37, Diman-kos ***@***.***>:

**Reply by Diman-kos** [reply]
Уведомления о платежах не поступают (
По логу apaсhe вообще не было никаких обращений к "incoming_payment.php".
Файл "incoming_payment_log.txt" тоже не создавался.

Вебхук продолжает пересоздаваться

**Reply by Diman-kos** [reply]
У меня не хватает ума понять в чём проблема )
Очень мудрёный API у точки!

**Reply by djpont** [reply]
Когда проходили авторизацию Точка Банка, поставили галочки на разрешения?
[Интеграцию](https://github.com/djpont/php-tochka-bank-telegram-notifications#%D0%B8%D0%BD%D1%82%D0%B5%D0%B3%D1%80%D0%B0%D1%86%D0%B8%D1%8F-%D1%81-%D1%82%D0%BE%D1%87%D0%BA%D0%B0-%D0%B1%D0%B0%D0%BD%D0%BA) с банком сделали по инструкции?

**Reply by Diman-kos** [reply]
Да, конечно.
Удалил все приложения в точке, проделал всё ещё раз по инструкции, обновил ключи доступа, дал разрешения (Приложению доступно: Получать остаток по счёту).

Ничего не изменилось (

Вебхук пересоздаётся заново каждый раз

**Reply by djpont** [reply]
Проверьте ещё, создался ли файл access_tokens.php
```
<?php
$refreshToken='A3d...DC8';
$accessToken='GaU...rJb';
$expiresTime='1688715161';
```

**Reply by Diman-kos** [reply]
Да, создался

**Reply by djpont** [reply]
Нашёл проблему, решилось добавлением принудительного сброса параметров перед новыми запросами.
Обновил, попробуйте теперь. Изменился `class.php` и поправил опечатку в `index.php`.

**Reply by Diman-kos** [reply]
Установил обновления.

Вывод скрипта:
"Авторизация прошла успешно.
Проверяю вебхук.
Вебхук требуется обновить.
Удаляю вебхук...
Создаю новый вебхук...
Новый вебхук создан!"

Вебхук продолжает пересоздаваться.
Но делает это гораздо быстрее ) Раньше скрипт минуту "думал", сейчас сразу же выдаёт

**Reply by Diman-kos** [reply]
Ждём поступлений на счёт для проверки работы уведомлений...

**Reply by djpont** [reply]
При обновлении странички снова пересоздаёт? Если да, попробуйте заново авторизоваться.
`?auth`

**Reply by Diman-kos** [reply]
> При обновлении странички снова пересоздаёт?

Да

> Если да, попробуйте заново авторизоваться

Попробовал.
Всё равно пересоздаёт

**Reply by Diman-kos** [reply]
Кажется понял в чём проблема....

`{
    "webhooksList": [
        "incomingPayment",
        "incomingSbpPayment"
    ],
    "url": ".../tochka/incoming_payment.php"
}`

`{
    "code": "400",
    "id": "e567d13e-1c67-4e64-9c31-4d8438d70096",
    "message": "Что-то пошло не так",
    "Errors": [
       {
            "errorCode": "Something going wrong",
            "message": "Failed to test webhook url accessibility",
            "url": "http://enter.tochka.com/doc/v2/redoc"
        }
    ]
}`

"Failed to test webhook url accessibility"
Кажется точка не может проверить доступ к файлу incoming_payment.php

Проверил что выдаёт incoming_payment.php. он выдаёт "ошибка 500"

PHP Fatal error:  Uncaught TypeError: Return value of TochkaBank::decodeJWT() must be of the type array, null returned in /var/www/..../tochka/class.php:121

Нужно сделать какой-то адекватный ответ на тестирование от точки!

**Reply by djpont** [reply]
А у вас https или http? Нужен обязательно https для работу хуков.
На сколько я вижу по строке 
`{ "webhooksList": [ "incomingPayment", "incomingSbpPayment" ], "url": ".../tochka/incoming_payment.php" }` хук создан успешно.
В url, естественно, должен быть ваш полный адрес до incoming_payment.php.

**Reply by djpont** [reply]
incoming_payment.php возвращает ошибку 500, если ему отправить пустой запрос.
Он принимает запросы только POST и в теле запроса должна быть JWT строка - в ней точка банк кодирует данные платежа.

**Reply by Diman-kos** [reply]
https есть.

`{ "webhooksList": [ "incomingPayment", "incomingSbpPayment" ], "url": ".../tochka/incoming_payment.php" }`
Это запрос отправляемый в точку. Ответом на него получено "Failed to test webhook url accessibility"

**Reply by Diman-kos** [reply]
> incoming_payment.php возвращает ошибку 500, если ему отправить пустой запрос. Он принимает запросы только POST и в теле запроса должна быть JWT строка - в ней точка банк кодирует данные платежа.

Да, я уже догадался про пустой запрос, но тем не менее ошибка "Failed to test webhook url accessibility" сохраняется

**Reply by djpont** [reply]
Как вариант, можете в `incoming_payment.php` добавить проверку, чтобы тело post запроса обязательно было. Тогда сможете открыть в браузере файл и проверить, что он что-то возвращает
```
<?php
include_once './class.php';
$postData = file_get_contents('php://input');
if (!empty($postData)) {
	$bank = new TochkaBank();
	$bank->incomingPaymentNotification($postData);
	exit;
}
echo 'Пустой запрос';
```

**Reply by Diman-kos** [reply]
Ура! Нашёл косяк!
Итак, проблема в проверке тестового ответа от точки class.php:420
```
if (!isset($data['SidePayer']) || !isset($data['SideRecipient'])) {
throw new Error('Ошибка обработки уведомления о платеже');
}
```

В тестовом уведомлении нет SidePayer и SideRecipient, поэтому вываливается ошибка.

Тестовое уведомление от точки:

`{"operationId": "A22001100263820100000533E625FCB3", "qrcId": "AS10006DPRTEFPFS9HJ9SQSDSVRHJD3L", "amount": "0.33", "payerMobileNumber": "+99991234567", "payerName": "\u0418\u0432\u0430\u043d \u0418\u0432\u0430\u043d\u043e\u0432\u0438\u0447 \u0418."}`

**Reply by Diman-kos** [reply]
Теперь вывод выглядит так:

Авторизация прошла успешно.
Проверяю вебхук.
Вебхук в порядке.

**Reply by Diman-kos** [reply]
Ждём платежи )

**Reply by djpont** [reply]
Отлично! А что за тестовое уведомление, если не секрет? Мне пришлось 10 руб в кассе пробивать для теста.
Уведомление отличается от того, что получаю я при переводах через СБП или при переводах от юрлиц. Возможно это перевод через qr-код? Странно, что, когда я создаю вебхук, то мне не приходит никаких проверочных уведомлений.

**Reply by Diman-kos** [reply]
А вот оно:
`{"operationId": "A22001100263820100000533E625FCB3", "qrcId": "AS10006DPRTEFPFS9HJ9SQSDSVRHJD3L", "amount": "0.33", "payerMobileNumber": "+99991234567", "payerName": "\u0418\u0432\u0430\u043d \u0418\u0432\u0430\u043d\u043e\u0432\u0438\u0447 \u0418."}`

Это шлёт точка при создании вебхука и проверяет ответ.

**Reply by djpont** [reply]
Кажется я понял. Согласно [документации](https://enter.tochka.com/doc/v2/redoc/section/Opisanie-metodov#Vebhuki) есть два типа платяжей: `incomingPayment` и `incomingSbpPayment`.
В организации, где я проверял работу уведомлений, они приходят с данными об отправителями и о получателе. Хотя оплата идёт через СБП.
Судя по вашему уведомлению, оно содержит id QR-кода, телефон, сумму и фио отправителя, что и должно быть при `incomingSbpPayment`.
Напишите, если уведомления действительно в таком виде будут приходить, то я добавлю поддержку таких уведомлений тоже.

**Reply by Diman-kos** [reply]
> Странно, что, когда я создаю вебхук, то мне не приходит никаких проверочных уведомлений

Да, странно.
Инфо из описания API точки:

> Для проверки доступности указанного вами хоста при создании или изменении вебхука мы отправим на ваш url по одному тестовому вебхуку на каждое из событий, на которые вы подписаны. Если в статусе ответа не придёт код HTTP 200, то вебхук не будет создан или изменён

**Reply by Diman-kos** [reply]
> Напишите, если уведомления действительно в таком виде будут приходить, то я добавлю поддержку таких уведомлений тоже.

Я не планировал использовать "incomingSbpPayment".
Но т.к. у вас по умолчанию скрипт подписывается на события "incomingSbpPayment", тестовое уведомление точка шлёт. И почему-то с id QR-кода

**Reply by djpont** [reply]
Добавил обработку incomingSbpPayment, теперь тестовое сообщение успешно приходит в Telegram, думаю и с настоящими платежами будет работать.

**Comment by Diman-kos** [comment]
Всё работает, уведомления приходят.
Но при входе в index.php
`{
    "message": "The access token is missing"
}`

Скрипт говорит:

Вебхук требуется обновить.
Удаляю вебхук...
Создаю новый вебхук...
`{
    "webhooksList": [
        "incomingPayment",
        "incomingSbpPayment"
    ],
    "url": "*****************************/tochka/incoming_payment.php"
}`

Tochka:
`{
    "message": "The access token is missing"
}`

Ошибка создания вебхука!

**Reply by djpont** [reply]
Спасибо за сообщение, постараюсь разобраться, почему проверка хуков не работает.
Отлично, что уведомления приходят.
